### PR TITLE
feat(executor): Wait for termination using pod watch for PNS and K8SAPI executors.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 

This PR introduces two improvements:

* If the PNS executor cannot get the main PID (because it is runAsNonRoot it uses a pod watch to determine if the `main` container terminated. This allows you to use PNS with `runAsNonRoot`, improving its security.
* Replaces the K8SAPI executor hard-loop with the same code, improving its scalability.

### CODE REVIEW

File: .github/workflows/ci-build.yaml [hunk 1]
```diff
@@ -32,18 +71,15 @@ jobs:
           go-version: "1.13.12"
       - name: Add bins to PATH
         run: |
-          echo "::add-path::/home/runner/go/bin"
-          echo "::add-path::/usr/local/bin"
+          echo /home/runner/go/bin >> $GITHUB_PATH
```
Comment (alexec) [inline .github/workflows/ci-build.yaml:74]: security fix for our actions

File: test/e2e/argo_server_test.go [hunk 2]
```diff
@@ -639,13 +639,10 @@ func (s *ArgoServerSuite) TestHintWhenWorkflowExists() {
 	s.e().POST("/api/v1/workflows/argo").
 		WithBytes([]byte((`{
   "workflow": {
-    "apiVersion": "argoproj.io/v1alpha1",
```
Comment (alexec) [inline test/e2e/argo_server_test.go:642]: this test should not have worked - I am confused

File: test/e2e/cli_test.go [hunk 3]
```diff
@@ -132,8 +132,6 @@ func (s *CLISuite) TestTokenArg() {
 	s.Run("ListWithGoodToken", func() {
 		s.Given().RunCli([]string{"list", "--user", "fake_token_user", "--token", goodToken}, func(t *testing.T, output string, err error) {
 			assert.NoError(t, err)
-			assert.Contains(t, output, "NAME")
```
Comment (alexec) [inline test/e2e/cli_test.go:135]: this test should not have worked either

File: test/e2e/fixtures/then.go [hunk 4]
```diff
@@ -49,6 +49,7 @@ func (t *Then) expectWorkflow(workflowName string, block func(t *testing.T, meta
 	if err != nil {
 		t.t.Fatal(err)
 	}
+	println(wf.Name, ":", wf.Status.Phase, wf.Status.Message)
```
Comment (alexec) [inline test/e2e/fixtures/then.go:52]: this is really useful

File: test/e2e/fixtures/when.go [hunk 5]
```diff
@@ -222,14 +222,18 @@ func (w *When) WaitForWorkflow(options ...interface{}) *When {
...
 			wf, ok := event.Object.(*wfv1.Workflow)
-			print(".")
 			if ok {
 				w.hydrateWorkflow(wf)
 				if condition(wf) {
 					println("Condition met after", time.Since(start).Truncate(time.Second).String())
 					w.wf = wf
 					return w
 				}
+				// once done the workflow is done, the condition can never be met
```
Comment (alexec) [inline test/e2e/fixtures/when.go:232]: this reduces the chances we'll get test suites that fail, but only after a long time

File: workflow/executor/common/wait/wait.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	for {
+		done, err := untilTerminatedAux(kubernetesInterface, namespace, containerID, listOptions)
+		if done {
+			return err
+		}
+	}
+}
+
+func untilTerminatedAux(kubernetesInterface kubernetes.Interface, namespace, containerID string, listOptions metav1.ListOptions) (bool, error) {
+	w, err := kubernetesInterface.CoreV1().Pods(namespace).Watch(listOptions)
```
Comment (alexec) [inline workflow/executor/common/wait/wait.go:27]: I looked at using `watch.Until` from the `go-client`, but it requires `resourceVersion` to work at all. There is `ListAndWatchUntil`, but this requires `list pods` but `workflow-role` does not have that permission.

File: workflow/executor/k8sapi/k8sapi.go [hunk 7]
```diff
@@ -60,8 +61,7 @@ func (k *K8sAPIExecutor) WaitInit() error {
 
 // Wait for the container to complete
 func (k *K8sAPIExecutor) Wait(containerID string) error {
-	log.Infof("Waiting for container %s to complete", containerID)
-	return k.client.waitForTermination(containerID, 0)
```
Comment (dtaniwaki) [inline workflow/executor/k8sapi/k8sapi.go:64]: Are we still using this method?

Comment (alexec) [reply workflow/executor/k8sapi/k8sapi.go:64]: do - I have deleted it

File: workflow/executor/common/wait/wait.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/argoproj/argo/workflow/executor/common"
+)
+
+func UntilTerminated(kubernetesInterface kubernetes.Interface, namespace, podName, containerID string) error {
+	log.Infof("Waiting for container %s to be terminated", containerID)
+	podInterface := kubernetesInterface.CoreV1().Pods(namespace)
```
Comment (sarabala1979) [inline workflow/executor/common/wait/wait.go:18]: I like this idea. It is safer than earlier

**Review by sarabala1979 — APPROVED** [review]
LGTM the implementations. @jessesuen Can you quickly look at the security point of view?

**Pull request merged**